Use `transaction.atomic` for various functions in actions.py.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Added commits to use `transaction.atomic` for some of the `do_...` functions in `actions.py`.
 <!-- How have you tested? -->

 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
@mateuszmandera can you review this one?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -4830,17 +4831,19 @@ def do_change_default_events_register_stream(
-            bot_owner_user_ids(user_profile),
+                bot_owner_user_ids(user_profile),
+            )
         )
```
Would it be better to construct the `event` object outside the `transaction.on_commit` block, so that it's just the `send_event` call that sits in there?

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -1403,6 +1403,7 @@ def send_user_email_update_event(user_profile: UserProfile) -> None:
 
+@transaction.atomic
 def do_change_user_delivery_email(user_profile: UserProfile, new_email: str) -> None:
     delete_user_profile_caches([user_profile])
```
Do we want this inside the `transaction.atomic` block? This hits memcached, so possibly we should think of this like we do of `send_event` calls, which hit tornado? On the other hand - `.save()` calls can often hit memcached too (due to post_save signal hooks), which is harder to avoid inside transactions and we're already doing that - so perhaps we should just give up on this and accept the fact that we will unavoidably hit memcached inside transactions?

timabbott replied:
Since it's safe to delete something from memcached for any reason at all (it's a cache), I think it's not a bug to do so inside a transaction block like this.

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -5344,6 +5344,7 @@ def update_scheduled_email_notifications_time(
     )
 
 
+@transaction.atomic
```
Wouldn't it make sense to put this only around this little bit of code?
```
    # TODO: Move these database actions into a transaction.atomic block.
    user_profile.save(update_fields=[setting_name])
    if setting_name in UserProfile.notification_setting_types:
        # Prior to all personal settings being managed by property_types,
        # these were only created for notification settings.
        #
        # TODO: Start creating these for all settings, and do a
        # backfilled=True migration.
        RealmAuditLog.objects.create(
            realm=user_profile.realm,
            event_type=RealmAuditLog.USER_SETTING_CHANGED,
            event_time=event_time,
            acting_user=acting_user,
            modified_user=user_profile,
            extra_data=orjson.dumps(
                {
                    RealmAuditLog.OLD_VALUE: old_value,
                    RealmAuditLog.NEW_VALUE: setting_value,
                    "property": setting_name,
                }
            ).decode(),
        )
```
(whatever we do, we can remove the TODO comment)

if we do, we should probably also just move up the `Draft.objects.filter(user_profile=user_profile).delete()` operation up here to be included in the transaction

sahil839 replied:
Do we not want `clear_scheduled_emails` and `update_scheduled_email_notifications_time` in the transaction ?

timabbott replied:
I think we probably do want those in the transaction, yeah, since they are coupled database updates.

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -4803,6 +4803,7 @@ def do_change_default_sending_stream(
         )
 
 
+@transaction.atomic
```
Similarly here, why don't we put this only around a part of the function, that deals with the db operation, and then we won't have to worry about adding `on_commit` etc. in other places?
```
    user_profile.save(update_fields=["default_events_register_stream"])
    event_time = timezone_now()
    RealmAuditLog.objects.create(
        realm=user_profile.realm,
        event_type=RealmAuditLog.USER_DEFAULT_REGISTER_STREAM_CHANGED,
        event_time=event_time,
        modified_user=user_profile,
        acting_user=acting_user,
        extra_data=orjson.dumps(
            {
                RealmAuditLog.OLD_VALUE: old_value,
                RealmAuditLog.NEW_VALUE: None if stream is None else stream.id,
            }
        ).decode(),
    )
```

sahil839 replied:
I am not sure if it was completely decided but [this] conversation seems to point out that we are fine with using `on_commit` for events.

timabbott replied:
I think the result will be the same with `transaction.on_commit` to send the event as with sending the event just after the end of the translation block.  In both cases, we're vulnerable to a race where the process fails (E.g. is killed by an external event) just after completing the database transaction but before the event is sent.

So it's really a question of which provides better readability. This is probably a question worth starting a thread on #backend about for visibility of discussion.

mateuszmandera reviewed (commented):
I've posted a couple of comments that mostly just have to do with putting `atomic()` around a small, relevant block in the function rather than decorating the whole function - the other commits seem to share analogical concerns, so I won't put repetitive comments on them :thinking: 

Ah, and I should point to my https://github.com/zulip/zulip/pull/20544 PR which does similar work for a couple other functions, so that we don't end up duplicating the work :grin:

sahil839 commented:
@timabbott I addressed the previous comments, added `durable=True` as per discussion in chat and also added a couple of new commits for more `do_..` functions.

Also, I have not added `durable=True` for `do_change_user_delivery_email`, `do_change_realm_plan_type` and `do_change_avatar_fields` for now as these functions are called from a transaction at some places and thus it generates runtime error. We cannot catch these in tests because -
> django.test.TestCase disables the durability check to allow testing durable atomic blocks in a transaction for performance reasons.
And have also moved these commits to end in case we do not want to merge these now.

timabbott commented:
I merged the first several commits as the series ending with 06cba4ae1f7818a0369f99f89b7bf7d3586e1eee.  

I think for those remaining functions, we should discuss in chat.zulip.org using `savepoint=False` instead with a comment noting that they are called as part of a large transaction.  It is definitely not correct to use `transaction.atomic` without some parameter in this setting.

sahil839 commented:
@timabbott added `savepoint=False` as per discussion.

timabbott commented:
Hmm, `do_set_realm_property` also calls `send_user_email_update_event`; what happens with a function with `transaction.on_commit` set on it when not inside a transaction?  It's possible we need to have the same merge migrate `do_set_realm_property` as well?

timabbott commented:
OK, great.

timabbott commented:
Merged, thanks @sahil839!
